echopop/acoustics.py

```diff
@@ -0,0 +1,452 @@
+
+    # Calculate mean TS for all hauls, KS-strata, and INPFC strata
+    # ---- By haul
+    sigma_bs_haul["TS_mean"] = to_dB(sigma_bs_haul["sigma_bs_mean"])
+    # ---- By KS stratum
+    sigma_bs_stratum_updated["TS_mean"] = to_dB(sigma_bs_stratum_updated["sigma_bs_mean"])
+
+    # Return output
+    return {"haul_mean_df": sigma_bs_haul, "strata_mean_df": sigma_bs_stratum_updated}
+
+
+def nasc_to_biomass(
+    input_dict: dict, analysis_dict: dict, configuration_dict: dict, settings_dict: dict
+) -> tuple[pd.DataFrame, pd.DataFrame]:
+    """
+    Converts integrated acoustic backscatter (NASC) into estimates of
+    areal number/biomass densities, total abundance, and total biomass
+
+    Parameters
+    ----------
+    input_dict: dict
+        A dictionary containing the loaded survey data.
+    analysis_dict: dict
+        A dictionary containing processed biological and transect data.
+    configuration_dict: dict
+        Dictionary that contains all of the `Survey`-object configurations found within
+        the `config` attribute.
+    settings_dict: dict
+        Dictionary that contains all of the analysis settings that detail specific algorithm
+        arguments and user-defined inputs.
+
+    Notes
+    -----
+    This function converts NASC into estimates of population-level metrics
+    (abundance, biomass, areal densities) stratified by transects, sex,
+    length-based strata, and age.
+    """
+
+    # Extract the necessary correct strata mean sigma_bs
+    sigma_bs_strata = analysis_dict["acoustics"]["sigma_bs"]["strata_mean_df"]
+
+    # Pull out the length-weight conversion for each stratum
+    length_weight_strata = analysis_dict["biology"]["weight"]["weight_stratum_df"]
+
+    # Get the name of the stratum column
+    stratum_col = settings_dict["transect"]["stratum_name"]
+
+    # Correct the acoustic survey transect intervals
+    nasc_interval_df = correct_transect_intervals(input_dict["acoustics"]["nasc_df"])
+
+    # Select the appropriate NASC column based on the inclusion or exclusion of age-1 fish
+    if settings_dict["transect"]["exclude_age1"]:
+        # ---- Drop the unused column
+        nasc_interval_df.drop("NASC_all_ages", axis=1, inplace=True)
+        # ---- Rename the used column
+        nasc_interval_df.rename(columns={"NASC_no_age1": "nasc"}, inplace=True)
+        # ---- Calculate age-1 NASC and weight proportions
+        age1_proportions = age1_metric_proportions(
+            input_dict["biology"]["distributions"],
+            analysis_dict["biology"]["proportions"],
+            configuration_dict["TS_length_regression_parameters"]["pacific_hake"],
+            settings_dict,
+        )
+        # ---- Calculate adult proportions
+        # -------- Initialize dataframe
+        adult_proportions = age1_proportions.copy()
+        # -------- Invert proportions
+        adult_proportions.iloc[:, np.arange(1, 4)] = (
+            1.0 - adult_proportions.iloc[:, np.arange(1, 4)]
+        )
```

Comment:
why do you need np.arange(1,4)? it seems a weird hard-coded thing. It seems that `age1_proportions` only has 3 columns from `age1_metric_proportions`?

Reply:
There are four columns: 1) stratum, 2) number proportion, 3) weight proportion, and 4) NASC proportion. Those are computed for all columns. Perhaps there is a more sophisticated way to map out applying the same simple subtraction to all three columns since `np.arange(1, 4)` at least yields the correct column numbers (`array([1, 2, 3]`) that should always remain constant regardless of the number of strata.

Reply:
I suggest that you use column names instead of indices, to make this more robust to errors.